Create PR for golangupx release

aminvakil opened the pull request:
Follow up to https://github.com/devops-works/scan-exporter/pull/27#issuecomment-3400297964.

aminvakil committed:
Create PR for golangupx release

aminvakil commented on .github/workflows/check-golangupx-release.yml:
```diff
@@ -101,10 +101,21 @@ jobs:
-          git push || echo "No changes to push"
+          git commit -m "chore: update Golang upx to version ${{ needs.check-golang-version.outputs.version }}"
+          git push --force -u origin golangupx_new_release
+          PR_NUMBER=$(gh pr list --head golangupx_new_release --json number --jq '.[0].number')
```
Here I check for existing PR, so it would not try to create another PR and fail in case there is one.

eze-kiel replied:
I think that having unique branch names (cf. my comment above) is a cleaner way to handle duplicates; this will avoid breaking stuff if the API changes or if the `gh` CLI changes its behavior

aminvakil replied:
Correct, but this is still needed in case PR has not been merged yet and workflow runs again (after a day for now).

eze-kiel replied:
This case can be avoided thanks to unique branches names (implemented in https://github.com/devops-works/scan-exporter/commit/2547e85ed2d75f0b14f0e54fffd6f9d880946a6c) and by doing the check before the whole Git stuff. For example, the flow could be:

- build the branch name based on the golang-upx version
- check if there's already a remote branch with the same name
  - if yes: exit (PR already exists)
  - if no: do the whole git magic (PR doesn't already exist)

This will also avoid having to use `--force` as we're no longer gonna push on an already existing branch (since we're exiting early)

aminvakil replied:
Correct, I'll change that.

aminvakil replied:
Fixed in https://github.com/devops-works/scan-exporter/pull/29/commits/5baebdd51e76b9192391a4716acebc2e45dfbaf1.

aminvakil commented on .github/workflows/check-golangupx-release.yml:
```diff
@@ -101,10 +101,21 @@ jobs:
+      - name: Create Pull Request
+        if: github.event_name != 'pull_request'
         run: |
+          set -x
```
This can be deleted later.

aminvakil committed:
Remove unnecessary event_name check

aminvakil committed:
Remove unnecessary pr listing

eze-kiel commented on .github/workflows/check-golangupx-release.yml:
```diff
@@ -101,10 +101,19 @@ jobs:
-          git commit -m "chore: update Golang upx to version ${{ needs.check-golang-version.outputs.version }}" || echo "No changes to commit"
-          git push || echo "No changes to push"
+          git commit -m "chore: update Golang upx to version ${{ needs.check-golang-version.outputs.version }}"
+          git push --force -u origin golangupx_new_release
```
I do not think the --force flag is required

aminvakil replied:
It fails without `--force` if branch exists with a commit on top of it (i.e. an open PR).

eze-kiel replied:
see my comment [here](https://github.com/devops-works/scan-exporter/pull/29#discussion_r2453047577); by checking early if the PR for the current version already exist, we can avoid pushing multiple times to an existing PR

eze-kiel commented on .github/workflows/check-golangupx-release.yml:
```diff
@@ -101,10 +101,19 @@ jobs:
+      - name: Create Pull Request
         run: |
+          set -x
+          git branch golangupx_new_release
```
please use dynamic branch name to avoid race conditions. For example:

```
golangupx_${{ needs.check-golang-version.outputs.version }}
```

aminvakil replied:
Fixed in https://github.com/devops-works/scan-exporter/pull/29/commits/2547e85ed2d75f0b14f0e54fffd6f9d880946a6c.

eze-kiel reviewed (changes requested):
Thanks for your PR!

A few changes are required, and also there are some issues if the original file:

 - your cronjob is `0 * * * *`, but the comment says "# Run every 24 hours" which is not true. Daily runs seems to be good enough so it may be worth updating the cron.
 - in light of recent supply chain attacks targeting dependencies in Github repo, it may be a good idea to pin the actions to a given SHA (on lines 17 and 82). Pinning the `golang-upx` SHA in the docker images as well as the `release_build.yml` is also a good idea, currently they can be pwned via supply chain attack.
 - I think the permissions are too wide, `packages: write` is probably not needed.

Thanks for taking time to improve the repo! Happy to help if needed

aminvakil committed:
Run every 24 hours for real

aminvakil commented:
> Thanks for your PR!
> 
> A few changes are required, and also there are some issues if the original file:
> 
> your cronjob is `0 * * * *`, but the comment says "# Run every 24 hours" which is not true. Daily runs seems to be good enough so it may be worth updating the cron.

Done, fixed, thanks!
> 
> in light of recent supply chain attacks targeting dependencies in Github repo, it may be a good idea to pin the actions to a given SHA (on lines 17 and 82). Pinning the `golang-upx` SHA in the docker images as well as the `release_build.yml` is also a good idea, currently they can be pwned via supply chain attack.

I think pros of using v4 of actions/checkout (an official github repository) is more than using a SHA. Because with using SHA we must keep up with all actions/checkout v4 releases which may contain a security fix.
> 
> I think the permissions are too wide, `packages: write` is probably not needed.
> 
https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#preventing-github-actions-from-creating-or-approving-pull-requests :
> By default, when you create a new repository in your personal account, workflows are not allowed to create or approve pull requests. If you create a new repository in an organization, the setting is inherited from what is configured in the organization settings.

This is needed for creating PRs by default, if devops-works organization is non-default, I can remove write permission.

Although write permission for packages is necessary though and I have removed it.

> 
> Thanks for taking time to improve the repo! Happy to help if needed

Thank you!

aminvakil committed:
Remove unnecessary write packages permission

aminvakil committed:
Use dynamic branch name

aminvakil committed:
Create PR if branch does not exist

eze-kiel reviewed (approved):
LGTM, thanks!
